UmpleParser/src/ParseUtilities_Code.ump

```diff
@@ -0,0 +1,606 @@
+    return getName().equals(aName) && isStatic();
+  }
+
+  public boolean isStatic()
+  {
+    return "STATIC".equals(getValue());
+  }
+
+  public boolean isValue(String compareTo)
+  {
+    return getValue().equals(compareTo);
+  }
+
+  public void remove(int index)
+  {
+    if (index < 0 || index >= subTokens.size())
+    {
+      return;
+    }
+    subTokens.remove(index);
+  }
+
+  public String toString()
+  {
+    StringBuffer answer = new StringBuffer();
+    toString(answer,null);
+    return answer.toString();
+  }
+
+  // Code refactored from Parser class to Token class
+  public StringBuffer toString(StringBuffer stringSoFar, Token parserRoot)
+  {
+    if ("START_TOKEN".equals(getValue()))
+    {
+      stringSoFar.append("[" + getName() + "]");
+    }
+    else if ("STATIC".equals(getValue()))
+    {
+      //ignore
+    }
+    else if (parserRoot == null || !this.equals(parserRoot))
+    {
+      stringSoFar.append("[" + getName() + ":" + getValue() + "]");
+    }
+
+    for(Token subToken : getSubTokens())
+    {
+      subToken.toString(stringSoFar,null);
+    }
+    return stringSoFar;
+  }
+  
+  // Return a copy of the current token with all its SubTokens
+  // Removes all SubTokens from the current Token and sets new name and value
+  public Token detach(String newName, String newValue){
+    Token nToken = this.shallowCopy();
+    setName(newName);
+    setValue(newValue);
+    ArrayList<Token> subTokenList = new ArrayList<Token>(subTokens);
+    for(Token t : subTokenList)
+    {
+      nToken.addSubToken(t);
+      this.removeSubToken(t);
+    }
+    return nToken;
+  }
+  
+  public Token copy(){
+    Token nToken = shallowCopy();
+    for (Token t : subTokens)
+    {
+      nToken.addSubToken(t.copy());    
+    }
+    return nToken;
+  }
+  
+  private Token shallowCopy()
+  {
+    Token nToken = new Token(this.getName()+"", this.getValue()+"");
+    if (getPosition() != null)
+    {
+      nToken.setPosition(getPosition().copy());      
+    }
+    if (getEndPosition() != null)
+    {
+      nToken.setEndPosition(getEndPosition().copy());
+    }
+    return nToken;
+  }
+
+}
+
+class Position
+{
+  depend java.nio.file.*;
+
+  public Position(int aLineNumber, int aCharacterOffset, int aOffset)
+  {
+    this(null, aLineNumber, aCharacterOffset, aOffset);
+  }
+
+  public Position copy()
+  {
+    return new Position(filename,lineNumber,characterOffset,offset);
+  }
+
+  public Position add(Position more)
+  {
+    String tfile = filename == null ? more.getFilename() : filename;
+    return new Position(filename, getLineNumber() + more.getLineNumber() - 1, getCharacterOffset() + more.getCharacterOffset(), getOffset() + more.getOffset());
+  }
+  
+  // extract the line number of the position
+  public int getLineNum() {
+    return getLineNumber();
+  }
+
+  public String toString()
+  {
+    return "[" + getLineNumber() + "," + getCharacterOffset() + "]";
+  }
+}
+
+class ParseResult
+{
+  before addErrorMessage {
```

Comment:
Why are these injections being done in the implementation details and not the interface? Just wondering on the justification.